lib/UBOTypeTransformPass.cpp

```diff
@@ -0,0 +1,520 @@
+  if (iter != remapped_types_.end()) {
+    return iter->second;
+  }
+
+  // Fix circular references.
+  if (!deferred_types_.insert(type).second) {
+    return type;
+  }
+
+  // Rebuild the types. Most types do not need handled here.
+  Type *remapped = type;
+  switch (type->getTypeID()) {
+  case Type::PointerTyID: {
+    PointerType *pointer = cast<PointerType>(type);
+    Type *pointee = MapType(pointer->getElementType(), M);
+    remapped = PointerType::get(pointee, pointer->getAddressSpace());
+    break;
+  }
+  case Type::StructTyID:
+    remapped = MapStructType(cast<StructType>(type), M);
+    break;
+  case Type::ArrayTyID: {
+    ArrayType *array = cast<ArrayType>(type);
+    Type *element = MapType(array->getElementType(), M);
+    remapped = ArrayType::get(element, array->getNumElements());
+    break;
+  }
+  case Type::FunctionTyID: {
+    FunctionType *function = cast<FunctionType>(type);
+    SmallVector<Type *, 8> arg_types;
+    for (auto *param : function->params()) {
+      arg_types.push_back(MapType(param, M));
+    }
+    remapped = FunctionType::get(MapType(function->getReturnType(), M),
+                                 arg_types, function->isVarArg());
+    break;
+  }
+  default:
+    break;
+  }
+
+  deferred_types_.erase(type);
+
+  auto result = remapped_types_.insert(std::make_pair(type, remapped));
+  if (remapped != type && result.second && !remapped->isFunctionTy()) {
+    // Record the type sizes from data layout to generate correct SPIRV-V
+    // information later.
+    const auto &DL = M.getDataLayout();
+    NamedMDNode *type_sizes_md =
+        M.getOrInsertNamedMetadata(clspv::RemappedTypeSizesMetadataName());
+    auto *i64 = Type::getInt32Ty(M.getContext());
+    Metadata *size_values[3];
+    size_values[0] = ConstantAsMetadata::get(
+        ConstantInt::get(i64, DL.getTypeSizeInBits(type)));
+    size_values[1] = ConstantAsMetadata::get(
+        ConstantInt::get(i64, DL.getTypeStoreSize(type)));
+    size_values[2] = ConstantAsMetadata::get(
+        ConstantInt::get(i64, DL.getTypeAllocSize(type)));
+    MDTuple *values_md = MDTuple::get(M.getContext(), size_values);
+    MDTuple *entry = MDTuple::get(
+        M.getContext(),
+        {ConstantAsMetadata::get(Constant::getNullValue(remapped)), values_md});
+    type_sizes_md->addOperand(entry);
+  }
+
+  return remapped;
+}
+
+StructType *UBOTypeTransformPass::MapStructType(StructType *struct_ty,
+                                                Module &M) {
+  bool changed = false;
+  SmallVector<Type *, 8> elements;
+  SmallVector<uint64_t, 8> offsets;
+  const auto *layout = M.getDataLayout().getStructLayout(struct_ty);
+  for (unsigned i = 0; i != struct_ty->getNumElements(); ++i) {
+    Type *element = struct_ty->getElementType(i);
+    uint64_t offset = layout->getElementOffset(i);
+    const auto *array = dyn_cast<ArrayType>(element);
+    if (array && array->getElementType()->isIntegerTy(8) && offset % 16 != 0) {
```

Comment:
The 16 is a bit magic here, but I'll allow it.
It's from the UBO layout requirements.